services/src/main/java/edu/unc/lib/dl/cdr/services/rest/modify/ExportCsvController.java

```diff
@@ -0,0 +1,261 @@
+                                                      HttpServletResponse response) throws IOException {
+
+        Map<String, Object> result = new HashMap<>();
+        result.put("action", "exportCsv");
+        result.put("username", request.getRemoteUser());
+
+        try {
+            if (pidString.equals("collections")) {
+                throw new FedoraException("Not allowed to download repository root");
+            }
+
+            PID pid = PIDs.get(pidString);
+
+            SearchRequest searchRequest = generateSearchRequest(request, searchStateFactory.createSearchState());
+            searchRequest.setRootPid(pid);
+            searchRequest.setApplyCutoffs(false);
+
+            SearchState searchState = searchRequest.getSearchState();
+            searchState.setResultFields(Arrays.asList(SearchFieldKeys.ID.name(), SearchFieldKeys.TITLE.name(),
+                    SearchFieldKeys.RESOURCE_TYPE.name(), SearchFieldKeys.ANCESTOR_IDS.name(),
+                    SearchFieldKeys.STATUS.name(), SearchFieldKeys.DATASTREAM.name(),
+                    SearchFieldKeys.ANCESTOR_PATH.name(), SearchFieldKeys.CONTENT_MODEL.name(),
+                    SearchFieldKeys.DATE_ADDED.name(), SearchFieldKeys.DATE_UPDATED.name(),
+                    SearchFieldKeys.LABEL.name(), SearchFieldKeys.CONTENT_STATUS.name()));
+            searchState.setSortType("export");
+            searchState.setRowsPerPage(searchSettings.maxPerPage);
+
+            BriefObjectMetadata container = queryLayer.addSelectedContainer(pid, searchState, false,
+                    searchRequest.getAccessGroups());
+            SearchResultResponse resultResponse = queryLayer.getSearchResults(searchRequest);
+
+            List<BriefObjectMetadata> objects = resultResponse.getResultList();
+            objects.add(0, container);
+
+            childrenCountService.addChildrenCounts(objects, searchRequest.getAccessGroups());
+
+            String filename = pid.getId().replace(":", "_") + ".csv";
+            response.addHeader("Content-Disposition", "attachment; filename=\"" + filename + "\"");
+            response.addHeader("Content-Type", "text/csv");
+
+            try (ServletOutputStream out = response.getOutputStream()) {
+                Writer writer = new BufferedWriter(new OutputStreamWriter(out, StandardCharsets.UTF_8));
+
+                try (CSVPrinter printer = getPrinter(writer)) {
+                    for (BriefObjectMetadata object : objects) {
+                        printObject(printer, object);
+                    }
+                }
+            }
+        } catch (FedoraException e) {
+            log.error("Error exporting CSV for {}. Not allowed to export collection root", pidString);
+            result.put("error", e.getMessage());
+            return new ResponseEntity<>(result, HttpStatus.FORBIDDEN);
+        } catch (NullPointerException e) {
```

Comment:
My understanding is this is to catch if `container` comes back as `null`, since that throws an NPE when calculating child size? It would be better/more accurate to do a null check right after assigning `container` a value and return a response with a 404 immediately.